src/ccache.cpp

```diff
@@ -1681,35 +1681,47 @@ from_cache(Context& ctx, FromCacheCallMode mode)
                                            : Statistic::preprocessed_cache_hit;
 }
 
-// Find the real compiler. We just search the PATH to find an executable of the
-// same name that isn't a link to ourselves.
+// Find the real compiler and put it into ctx.orig_args[0]. We just search the
+// PATH to find an executable of the same name that isn't a link to ourselves.
 static void
-find_compiler(Context& ctx, const char* const* argv)
+find_compiler(Context& ctx)
 {
-  // We might be being invoked like "ccache gcc -c foo.c".
-  std::string base(Util::base_name(argv[0]));
-  if (Util::same_program_name(base, CCACHE_NAME)) {
-    ctx.orig_args.pop_front();
-    if (Util::is_full_path(ctx.orig_args[0])) {
-      return;
-    }
-    base = std::string(Util::base_name(ctx.orig_args[0]));
-  }
+  // We might be being invoked like:
+  // 1) "/path/to/ccache relative/gcc -c foo.c"
+  // 2) "/path/to/ccache /absolute/path/gcc -c foo.c"
+  // 3) "/path/to/ccache gcc -c foo.c"
+  // 4) "/path/to/gcc -c foo.c" (where gcc is a link to ccache)
+
+  // Case 1, 2 or 3
+  const bool first_param_is_ccache =
+    Util::same_program_name(Util::base_name(ctx.orig_args[0]), CCACHE_NAME);
 
   // Support user override of the compiler.
+  std::string compiler;
   if (!ctx.config.compiler().empty()) {
-    base = ctx.config.compiler();
+    compiler = ctx.config.compiler();
+  } else {
+    compiler = ctx.orig_args[first_param_is_ccache ? 1 : 0];
   }
 
-  std::string compiler = find_executable(ctx, base, CCACHE_NAME);
-  if (compiler.empty()) {
-    throw Fatal("Could not find compiler \"{}\" in PATH", base);
+  // Resolve in case of 4 (not 1, 2, 3) or 3 (not 1, 2, 4)
```

Comment:
Yes sorry about that. I'm a big fan of "if you need to write a comment then you failed". And here I failed. Horribly. I wrote "not 1, 2, 3" since that is what is technically in the if clause. I'll introduce a variable instead. Maybe that will be better.